lib/promscrape/discovery/kubernetes/endpointslices.go

```diff
@@ -0,0 +1,214 @@
+			}
+
+		}
+	}
+
+	// Append labels for skipped ports on seen pods.
+	portSeen := func(port int, ports []int) bool {
+		for _, p := range ports {
+			if p == port {
+				return true
+			}
+		}
+		return false
+	}
+	for p, ports := range podPortsSeen {
+		for _, c := range p.Spec.Containers {
+			for _, cp := range c.Ports {
+				if portSeen(cp.ContainerPort, ports) {
+					continue
+				}
+				addr := discoveryutils.JoinHostPort(p.Status.PodIP, cp.ContainerPort)
+				m := map[string]string{
+					"__address__": addr,
+				}
+				p.appendCommonLabels(m)
+				p.appendContainerLabels(m, c, &cp)
+				ms = append(ms, m)
+			}
+		}
+	}
+	return ms
+
+}
+
+// getEndpointSliceLabelsForAddressAndPort gets labels for endpointSlice
+// from  address, Endpoint and EndpointPort
+// enriches labels with TargetRef
+// pod appended to seen Ports
+// if TargetRef matches
+func getEndpointSliceLabelsForAddressAndPort(podPortsSeen map[*Pod][]int, addr string, eps *EndpointSlice, ea Endpoint, epp EndpointPort, p *Pod, svc *Service) map[string]string {
+	m := getEndpointSliceLabels(eps, addr, ea, epp)
+	if svc != nil {
+		svc.appendCommonLabels(m)
+	}
+	if ea.TargetRef.Kind != "Pod" || p == nil {
+		return m
+	}
+	p.appendCommonLabels(m)
+	for _, c := range p.Spec.Containers {
+		for _, cp := range c.Ports {
+			if cp.ContainerPort == epp.Port {
+				p.appendContainerLabels(m, c, &cp)
+				podPortsSeen[p] = append(podPortsSeen[p], cp.ContainerPort)
+				break
+			}
+		}
+	}
+
+	return m
+}
+
+// //getEndpointSliceLabels builds labels for given EndpointSlice
+func getEndpointSliceLabels(eps *EndpointSlice, addr string, ea Endpoint, epp EndpointPort) map[string]string {
+
+	var status bool
+	if ea.Conditions.Ready != nil && *ea.Conditions.Ready {
+		status = true
+	}
+	addr = discoveryutils.JoinHostPort(addr, epp.Port)
+	m := map[string]string{
+		"__address__":                                               addr,
+		"__meta_kubernetes_namespace":                               eps.Metadata.Namespace,
+		"__meta_kubernetes_endpointslice_name":                      eps.Metadata.Name,
+		"__meta_kubernetes_endpointslice_address_type":              eps.AddressType,
+		"__meta_kubernetes_endpointslice_endpoint_conditions_ready": strconv.FormatBool(status),
+		"__meta_kubernetes_endpointslice_port_name":                 epp.Name,
+		"__meta_kubernetes_endpointslice_port_protocol":             epp.Protocol,
+		"__meta_kubernetes_endpointslice_port_app_protocol":         epp.AppProtocol,
+		"__meta_kubernetes_endpointslice_port":                      strconv.FormatUint(uint64(epp.Port), 10),
+	}
+	if ea.TargetRef.Kind != "" {
+		m["__meta_kubernetes_endpointslice_address_target_kind"] = ea.TargetRef.Kind
+		m["__meta_kubernetes_endpointslice_address_target_name"] = ea.TargetRef.Name
+	}
+	if ea.Hostname != "" {
+		m["__meta_kubernetes_endpointslice_endpointslice_endpoint_hostname"] = ea.Hostname
```

Comment:
It looks like the second `endpointslice` isn't needed in this label

Reply:
Could you also make sure that all the generated labels are checked in the test?

Reply:
Fixed. Its already checked, hm, maybe i missed smthng.